### PR TITLE
Enable tracking custom hparam metrics for TensorBoardLogger

### PR DESCRIPTION
**Pull request by its-dron**
To compare metrics in the HPARAM tab of TensorBoard, `TensorBoardLogger` requires the metrics to track be specified during the initial call to `log_hyperparameters`. 

This change adds a method to expose the methods to be tracked from the `LightningModule`, and passes them to the logger if applicable. This allows more features of TensorBoard to be used, as shown below.

Screen Shot 2021-08-12 at 2 31 01 PM

See [here](https://pytorch-lightning.readthedocs.io/en/latest/extensions/logging.html#logging-hyperparameters) for more details on hparam metrics and `TensorBoardLogger`.

### CODE REVIEW

**Comment by ashleve** [comment]
Thanks for the proposition!
Seems useful when logging with Tensorboard logger, however, it convolutes code a bit. It also forces users to add `get_hparam_metrics` method to every model they write.

The main goal of this template is to be as clean as possible to make it easier to understand the libraries. As such, would prefer not to add the change specific to only one logger to keep the template simple

**Comment by its-dron** [comment]
I will point out that the `get_hparam_metrics` method is optional, and only needed if you want to use this feature. If omitted, functionality will be exactly the same as the current code. 

But keeping a simple template is understandable.

**Comment by ashleve** [comment]
@its-dron My bad, you're right it's optional.

Is it possible to log those metric names only on initialization? Otherwise, this kind of logger specific functionality could be exported into callback so it's not dependent on logger code

**Comment by its-dron** [comment]
Yes, they have to be logged the first time `TensorBoardLogger.log_hyperparameters` is called. Here is a minimal example which shows that it doesn't work:

```
from pytorch_lightning.loggers import TensorBoardLogger
logger = TensorBoardLogger("tb_logs", name="my_model")
params = {'size': 100, 'dims': [10,100,20], 'model': {'height': 10, 'width': 20}}
logger.log_hyperparams(params)
metrics = {'test': 0, 'val': 0}
logger.log_hyperparams(params, metrics)
```

However, you can log other things before you log the hparams, as long as you disable `default_hp_metric`.

```
from pytorch_lightning.loggers import TensorBoardLogger
logger = TensorBoardLogger("tb_logs", name="my_model_no_default", default_hp_metric=False)
params = {'size': 100, 'dims': [10,100,20], 'model': {'height': 10, 'width': 20}}
metrics = {'test': 0, 'val': 0}
logger.log_metrics(metrics)             # adds scalars, no hparams
logger.log_hyperparams(params)          # doesn't do anything?
logger.log_hyperparams(params, metrics) # adds hparams correctly! but redundantly logs scalars
```

**Comment by nils-werner** [comment]
You probably can also use callbacks for this, with something similar to

```python
# src/callbacks/loghyperparameters.py
from pytorch_lightning import Callback, Trainer
from pytorch_lightning.utilities import rank_zero_only


class LogHyperParameters(Callback):
    def __init__(self, params, metrics):
        self.params = params
        self.metrics = metrics

    @rank_zero_only
    def on_train_end(self, trainer, pl_module):
        print(self.params)
        print(trainer.model.hparams)
        print(trainer.callback_metrics)
        # trainer.logger.log_hyperparameters(...)
```

```yaml
#config/callbacks/default.yaml

# ...

hparam_logging:
  _target_: src.callbacks.loghyperparameters.LogHyperParameters
  params: ${model}
  metrics:
    - "val/acc"
```
